### PR TITLE
LLVM-Passes: add pointer authentication to Swift's function merge pass.

### PR DESCRIPTION
**Pull request by eeckstein**
If during merging a function pointer is passed as a parameter to the merged function, it needs to be signed on arm64e.

rdar://problem/66797689

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/LLVMPasses/LLVMMergeFunctions.cpp [hunk 1]
```diff
@@ -316,6 +325,15 @@ class SwiftMergeFunctions : public ModulePass {
       CurrentInst = &*std::next(CurrentInst->getIterator());
     }
 
+    /// Returns true if the operand \p OpIdx of the current instruction is the
+    /// callee of a call, which needs to be signed if passed as a parameter.
```
[the remark this answers is not in the document]
Comment (eeckstein) [reply lib/LLVMPasses/LLVMMergeFunctions.cpp:329]: Intrinsics are excluded from being parameterized, anyway.

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/LLVMPasses/LLVMMergeFunctions.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjmccall) [reply lib/LLVMPasses/LLVMMergeFunctions.cpp:341]: Okay, if the parameterization only happens with function calls and not other uses of functions, then yeah, this isn't a problem.

File: lib/LLVMPasses/LLVMMergeFunctions.cpp [hunk 2]
```diff
@@ -351,10 +382,28 @@ class SwiftMergeFunctions : public ModulePass {
       }
       return true;
     }
+    
+    /// Computes the discriminator for pointer signing.
+    void computeDiscriminator(LLVMContext &Context) {
+      assert(needsPointerSigning);
+      assert(!discriminator);
+      
+      /// Get a hash from the concatenated function names.
```
[the remark this answers is not in the document]
Comment (rjmccall) [reply lib/LLVMPasses/LLVMMergeFunctions.cpp:402]: Thanks

File: lib/LLVMPasses/LLVMMergeFunctions.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjmccall) [reply lib/LLVMPasses/LLVMMergeFunctions.cpp:341]: Oh, I see, parameterization skips all uses of intrinsics.

**Review by rjmccall — APPROVED** [review]
LGTM

**Pull request merged**